### ISSUE TITLE
Cross-device tensor copy has inconsistent/weird/broken stream behavior

### ISSUE THREAD

**Issue by wickedfoo** [torch/cutorch#212]
Which stream the cross-device copy uses has has different rules at present depending upon whether it is contig or non-contig, which is rather non-obvious to the user. This causes problems when you attempt to use non-default streams, which I'm getting very deep into at the moment.

The code:
```
  bool succ =
        THCudaTensor_pointwiseApply2(state, dst, src, CopyOp<float>());
```

can actually not use a stream on either the src or dest device (just whatever the current device is), since pointwiseApply is getting its stream via THCState, whereas that is not changed when you call cudaSetDevice.

It is unclear to me if cudaMemcpyAsync allows you to use a stream that has nothing to do with the src or dst devices. Right now, you can cutorch.setDevice() to a device that has nothing to do with the src or dst devices, and it's kind of up in the air as to what that means. Perhaps that should be an error.

I propose changing the tensor copy code to obey the following rules:

===
1. All copies (contig or not) are made from the src tensor device (regardless of what cutorch.getDevice returns at the point of :copy() invocation), using whatever stream the current stream index is on the src device. This removes the "optimization" for non-contig copies  but this makes it more sensible I think, so you always know what stream/device the copy is ordered with respect to.
Looking at nvprof/nvvp, if you issue the copy from a stream off of the src device, the CUDA API creates an internal stream for use on the src device, and adds additional synchronization. If you issue the copy from a src device stream, the p2p copy always occurs on the stream you intend.

2. If a copy is made on the default stream (THCState_getCurrentStreamIndex(state) == 0, regardless of what the current cutorch.getDevice returns at the point of :copy() invocation), then the default stream on both the src and dst devices will have a full barrier:

  -src stream 0 syncs to dst stream 0 (no one should be writing into dst before starting)
  -copy occurs on src device, stream 0
  -dst stream 0 syncs to src stream 0 (the copy should be complete before dst continues)

  This is slightly weaker than my interpretation of cudaMemcpyAsync in the NULL stream cross-device according to http://docs.nvidia.com/cuda/cuda-c-programming-guide/index.html#peer-to-peer-memory-copy, in that only the default stream involved on each device participates in a barrier, rather than all streams. This is basically the barrier that exists in the current code (though my reading of the CUDA documentation leads me to believe that the barrier is not needed for cudaMemcpyAsync on the default stream as added in https://github.com/torch/cutorch/commit/6ca69073563b1d20c16a375086cba072d42b3115). We'd just always issue the barrier to be safe though. I've sent some questions to NVIDIA on this and related issues.

3. If a copy is made off the default stream (THCState_getCurrentStreamIndex(state) > 0, regardless of what the current cutorch.getDevice returns), no two-way synchronization occurs. The only synchronization that occurs is with respect to that stream on the source device.

  The rationale here is that you may want to synchronize a different stream on the dst device to the src copy, so you have to issue the cutorch.streamWaitForMultiDevice yourself. e.g., copy occurs on (dev 2, stream 3) -> dev 3, but you want to then operate on that copy on dev 3 in dev 3's stream 6. Right now, you can't express that at all with the current API from Lua land, and there may be no relationship between (dev A, stream index N) and (dev B, stream index N) that you want; it would just be pointless synchronization.

  If you're running off of the default stream, you assume these synchronization responsibilities.

Agree? Disagree?

**Comment by wickedfoo** [comment]
(3) should read THCState_getCurrentStreamIndex(state) > 0, I edited it FYI.

**Comment by dominikgrewe** [comment]
You make some good points. The current synchronization strategy is really too restrictive. Given that a user can specify the synchronization with `cutorch.streamWaitForMultiDevice` I'm happy to drop the automatic synchronization.

I'm not so sure about the special treatment of the default stream. CUDA does it but even they seem to be moving away from it (CUDA 7 has an option that disables the special treatment of the default stream). I never really liked it, because it means your code isn't very composable. The behaviour changes depending on the context it's being executed in.
Are you worried that removing the automatic synchronization on the default stream will break existing code? Or do you just want to mirror what `cudaMemcpyAsync` is doing? I wonder if CUDA 7 allows you to disable the automatic synchronization when using the default stream.

**Comment by wickedfoo** [comment]
I got back some answers from NVIDIA.

Regarding the current device (as per `cudaGetDevice`) or stream into which cudaMemcpyAsync is issued, the synchronization will still work with respect to that stream, even if the current device or current stream has nothing to do with the two devices participating in the copy. The driver creates new streams as needed on any devices participating in the copy, and inserts in events to order everything with respect to the stream you are using.

Regarding special treatment of the default stream, there are two issues under consideration: what the default stream means based on  `--default-stream null` versus `--default-stream per-thread`, and how synchronous or asynchronous calls launched in that stream or other streams are handled.

I don't think the "default stream" concept is going away in CUDA. This is the stream that is used for ordering with respect to synchronous calls that don't take any stream arguments, like `cudaMemcpy()`. If you wanted to get rid of any "default stream", that would change the API pretty radically: every kernel launch would have to specify some stream, and every synchronous call (like cudaMemcpy) would have to do so as well. My understanding is what NVIDIA wants to deprecate is the `--default-stream null` behavior, which is what you get by default if you specify no #defines or nvcc arguments. This is what we're operating under right now.

`--default-stream null`, what we're using right now, means that the default stream (`cudaStream_t == 0`) is shared between all threads in the current process. Any work issued to that stream (either an asynchronous call like a kernel launch or a `cudaMemcpyAsync`, or a synchronous call like `cudaMemcpy`) will wait for the completion of all other work issued on all other streams (and in the default stream itself) before completion. The only streams that it doesn't wait for are those created with `cudaStreamCreateWithFlags(p, cudaStreamNonBlocking)`. `cutorch.reserveStreams(n)` creates streams without `cudaStreamNonBlocking`. This means that if you have kernels launched in the default stream at present, in any CPU thread, no new work will be executed in one of these user-created streams until it is complete. `cudaMemcpy()` (synchronous) will wait for work to complete in any user created streams as well before the copy.

`--default-stream per-thread` is the other option. There are two implications of this: unlike the above option, the default stream is not shared between all threads in the current process, and there is nothing special about this stream, insofar that it doesn't block work on other streams from completing.

As long as CUDA calls are only ever issued from any other thread in the Torch process, I think it is safe for us to switch to `--default-stream per-thread` as a nvcc argument. `cudaMemcpy()` (synchronous) will still be ordered with respect to this thread. This will ensure that a mix of calls made to the default stream and to user-created streams will potentially run in parallel.

However, what would involve a lot more careful work would be switching to `cudaStreamNonBlocking` for user-created streams. My understanding is If you do this, synchronous calls like `cudaMemcpy` are no longer synchronous with respect to whatever stream is producing the value. Every place (in all of cutorch, or any sub-library)` where we copy a value back to the host, where the value is being produced on a non-default stream, we'd have to issue a `cudaStreamSynchronize(stream)` before the `cudaMemcpy`.

So with this information, I don't think I'd change anything about the proposal (I have the code implemented locally, still testing it out since I'm using all kinds of automatically generated stream assignments in what I'm doing).

Stream safety can be rather subtle, and I think if you continue using the default stream, we should strive for safe synchronization between two devices, at the expense of possible performance or parallelism. No one else should be writing into the dst when you're copying, and the dst should wait for the src copy to complete. If you opt into non-default streams, you're taking the gun into your own hands.

Maybe the only thing we could change is performing the copy in an ordered manner with respect to whatever stream/device is current to make the :copy call, which the current code kind of sort of does in a buggy way. If that stream/device has nothing to do with the src or dest devices, then it's easy for a user to introduce race conditions unless they understand this behavior, and they'd have to add the synchronization themselves on the src and dest devices. That's more difficult to understand.

**Comment by dominikgrewe** [comment]
Thanks for the detailed explanation. The more I think about CUDA's default stream, the more I dislike it. It's very complicated to reason about when working with multiple threads and/or multiple streams. Whereas the behaviour of "async streams" (created with `cudaStreamCreate*`) is very predictable imo, as long as you avoid functions that don't take stream arguments, such as `cudaMemcpy`.

I'm wondering if we should not use the CUDA default stream at all. I've create a separate issue for that, as it's somewhat orthogonal to this one (see #217).

Regarding the original issue I'm fine with your solution.

**Comment by soumith** [comment]
I believe it's fixed now, thanks to @wickedfoo and thanks to NVIDIA.
